File: backend/doppelganger/store.py
"""In-memory vector store for the PoC. One instance per subject/twin.

Deliberately tiny (numpy-free cosine) so the PoC runs with no infra. For
prod, swap this class for pgvector/Qdrant behind the same method surface:
`add`, `search`, `all`, `count`.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from .embeddings import EmbeddingProvider, cosine
from .models import MemoryItem


@dataclass
class MemoryStore:
    embedder: EmbeddingProvider
    _items: dict[str, MemoryItem] = field(default_factory=dict)
    _vectors: dict[str, list[float]] = field(default_factory=dict)

    async def add(self, items: list[MemoryItem]) -> int:
        """Add items, skipping ids we already have (idempotent polling)."""
        fresh = [it for it in items if it.id not in self._items]
        if not fresh:
            return 0
        vecs = await self.embedder.embed([it.embedding_text() for it in fresh])
        for it, v in zip(fresh, vecs):
            self._items[it.id] = it
            self._vectors[it.id] = v
        return len(fresh)

    async def search(self, query: str, top_k: int) -> list[tuple[MemoryItem, float]]:
        if not self._items:
            return []
        qv = (await self.embedder.embed([query]))[0]
        scored = [
            (self._items[i], cosine(qv, v)) for i, v in self._vectors.items()
        ]
        scored.sort(key=lambda t: t[1], reverse=True)
        return scored[:top_k]

    def all(self) -> list[MemoryItem]:
        return list(self._items.values())

    def count(self) -> int:
        return len(self._items)

File: app/lib/engine/store.dart
import 'embeddings.dart';
import 'models.dart';

/// In-memory vector store — one per twin, lives for the app session.
class MemoryStore {
  final _embedder = HashEmbedding();
  final _items = <String, MemoryItem>{};
  final _vectors = <String, List<double>>{};

  int get count => _items.length;
  List<MemoryItem> all() => _items.values.toList();

  /// Idempotent: re-adding the same id is a no-op (safe to re-poll).
  int add(List<MemoryItem> items) {
    var added = 0;
    for (final it in items) {
      if (_items.containsKey(it.id)) continue;
      _items[it.id] = it;
      _vectors[it.id] = _embedder.embed(it.embeddingText());
      added++;
    }
    return added;
  }

  List<MapEntry<MemoryItem, double>> search(String query, int topK) {
    if (_items.isEmpty) return [];
    final qv = _embedder.embed(query);
    final scored = _vectors.entries
        .map((e) => MapEntry(_items[e.key]!, cosine(qv, e.value)))
        .toList()
      ..sort((a, b) => b.value.compareTo(a.value));
    return scored.take(topK).toList();
  }
}
